File: barbara-v3/src/services/signalwire.ts
/**
 * SignalWire Service
 * Handle outbound call placement via SignalWire REST API
 */

import fetch from 'node-fetch';
import { logger } from '../utils/logger.js';

const SW_PROJECT_ID = process.env.SIGNALWIRE_PROJECT_ID;
const SW_TOKEN = process.env.SIGNALWIRE_API_TOKEN;
const SW_SPACE = process.env.SIGNALWIRE_SPACE;

export interface OutboundCallParams {
  to: string;           // E.164 phone number
  from: string;         // SignalWire number
  webhookUrl: string;   // Barbara V3 webhook URL
  leadId?: string;
  brokerId?: string;
}

export interface SendSmsParams {
  to: string;
  from: string;
  body: string;
  statusCallback?: string;
  metadata?: Record<string, any>;
}

interface SignalWireCredentials {
  projectId: string;
  token: string;
  space: string;
}

function requireCredentials(): SignalWireCredentials {
  if (!SW_PROJECT_ID || !SW_TOKEN || !SW_SPACE) {
    throw new Error('SignalWire credentials not configured');
  }

  return {
    projectId: SW_PROJECT_ID,
    token: SW_TOKEN,
    space: SW_SPACE
  };
}

/**
 * Place an outbound call via SignalWire REST API
 * 
 * @param params - Call parameters
 * @returns SignalWire Call SID
 */
export async function placeOutboundCall(params: OutboundCallParams): Promise<string> {
  const { projectId, token, space } = requireCredentials();

  const { to, from, webhookUrl, leadId, brokerId } = params;
  
  // Build webhook URL with context
  let fullWebhookUrl = `${webhookUrl}?direction=outbound&from=${from}&to=${to}`;
  if (leadId) fullWebhookUrl += `&lead_id=${leadId}`;
  if (brokerId) fullWebhookUrl += `&broker_id=${brokerId}`;

  const url = `https://${space}/api/laml/2010-04-01/Accounts/${projectId}/Calls.json`;
  
  const body = new URLSearchParams({
    To: to,
    From: from,
    Url: fullWebhookUrl
  });

  logger.info(`📞 Placing outbound call: ${from} → ${to}`);

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Authorization': `Basic ${Buffer.from(`${projectId}:${token}`).toString('base64')}`
    },
    body: body.toString()
  });

  if (!response.ok) {
    const errorText = await response.text();
    logger.error(`SignalWire call failed: ${response.status} ${errorText}`);
    throw new Error(`Failed to place call: ${response.status}`);
  }

  const data: any = await response.json();
  const callSid = data.sid;
  
  logger.info(`✅ Call placed successfully: ${callSid}`);
  
  return callSid;
}

/**
 * Send an outbound SMS via SignalWire REST API
 */
export async function sendSmsMessage(params: SendSmsParams): Promise<any> {
  const { projectId, token, space } = requireCredentials();
  const { to, from, body, statusCallback } = params;

  const url = `https://${space}/api/laml/2010-04-01/Accounts/${projectId}/Messages.json`;

  const payload = new URLSearchParams({
    To: to,
    From: from,
    Body: body
  });

  if (statusCallback) {
    payload.append('StatusCallback', statusCallback);
  }

  logger.info(`💬 Sending SMS: ${from} → ${to}`);

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Authorization': `Basic ${Buffer.from(`${projectId}:${token}`).toString('base64')}`
    },
    body: payload.toString()
  });

  if (!response.ok) {
    const errorText = await response.text();
    logger.error(`SignalWire SMS failed: ${response.status} ${errorText}`);
    throw new Error(`Failed to send SMS: ${response.status}`);
  }

  const json = await response.json();
  return json;
}

File: livekit-agent/services/signalwire.py
"""SignalWire client for outbound calls"""
import logging
import base64
import httpx
from typing import Dict, Any, Optional
from config import Config

logger = logging.getLogger(__name__)


class SignalWireClient:
    """Client for SignalWire REST API"""
    
    def __init__(self):
        self.project_id = Config.SIGNALWIRE_PROJECT_ID
        self.token = Config.SIGNALWIRE_TOKEN
        self.space = Config.SIGNALWIRE_SPACE
        
        if not all([self.project_id, self.token, self.space]):
            logger.warning("⚠️ SignalWire credentials not fully configured")
        
        self.base_url = f"https://{self.space}/api/laml/2010-04-01/Accounts/{self.project_id}"
    
    def _get_auth_header(self) -> str:
        """Generate Basic Auth header"""
        auth_string = f"{self.project_id}:{self.token}"
        auth_bytes = auth_string.encode('utf-8')
        auth_b64 = base64.b64encode(auth_bytes).decode('utf-8')
        return f"Basic {auth_b64}"
    
    async def create_call(
        self,
        to: str,
        from_number: str,
        url: str,
        status_callback: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Create an outbound call via SignalWire REST API
        
        Args:
            to: Destination phone number (E.164 format)
            from_number: Source phone number (E.164 format)
            url: SWML URL to execute when call answers
            status_callback: Optional webhook for call status updates
        
        Returns:
            Dict with call details including 'sid'
        """
        if not all([self.project_id, self.token, self.space]):
            raise ValueError("SignalWire credentials not configured")
        
        api_url = f"{self.base_url}/Calls.json"
        
        # Build form data
        form_data = {
            "To": to,
            "From": from_number,
            "Url": url
        }
        
        if status_callback:
            form_data["StatusCallback"] = status_callback
        
        headers = {
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json",
            "Authorization": self._get_auth_header()
        }
        
        logger.info(f"📞 Creating SignalWire call: {from_number} → {to}")
        
        async with httpx.AsyncClient() as client:
            response = await client.post(
                api_url,
                headers=headers,
                data=form_data,
                timeout=30.0
            )
            
            if not response.is_success:
                error_text = response.text
                logger.error(f"❌ SignalWire call failed: {response.status_code} {error_text}")
                raise Exception(f"SignalWire API error: {response.status_code}")
            
            data = response.json()
            call_sid = data.get("sid")
            
            logger.info(f"✅ SignalWire call created: {call_sid}")
            
            return {
                "sid": call_sid,
                "status": data.get("status", "queued"),
                "to": to,
                "from": from_number
            }

